### PR TITLE
ENH: Infer integer-na in infer_dtype

### LINKED ISSUE #27283: ENH: infer_dtype should infer integer-na
xref #26272 
xref #27267 

As the first step of moving towards integer-na dtypes as the primary integer type, we need to teach ``infer_dtype`` that ``integer-na`` is a valid inferred type, right now

```
In [1]: from pandas.api.types import infer_dtype                                                                                                                                                                                                             

In [3]: infer_dtype([2, 3,4], skipna=False)                                                                                                                                                                                                                  
Out[3]: 'integer'

In [4]: infer_dtype([2, 3, 4, np.nan], skipna=False)                                                                                                                                                                                                         
Out[4]: 'mixed-integer-float'

In [5]: infer_dtype([2, 3, 4.2, np.nan], skipna=False)                                                                                                                                                                                                       
Out[5]: 'mixed-integer-float'
```

[4] could return 'integer-na' to indicate that we might want to infer ``Int64`` dtype and is distinct from the inferred type of [5] which must become ``float64``.

This will allow us to then support changing integer columns when we add nulls to Int64 rather than coerce to float64; this is pretty common in indexing setting operations. 

Secondly we can then enable ``.to_numeric`` to infer to integer-na (or unsigned-na) and the corresponding dtypes (#26272).

Finally we could support coercion of ``object`` dtypes from integers and nulls to coerce to Int64 (#27267 for ``.explode()`` and ``.infer_objects()``

This issue itself only is a very minor user facing change (infer_dtype itself).

### PR DESCRIPTION
**Pull request by jiangyue12392**
- [x] closes #27283
- [x] tests added/passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

As mentioned in #27335, this PR serves as an infrastructure for future changes in the default int type

### CODE REVIEW

**Review by gfyoung — CHANGES_REQUESTED** [review]
A test or two would be good to have here I imagine.

**Comment by jiangyue12392** [comment]
@gfyoung I am not sure where I should add the test since it has no interface change to users.

**Comment by jiangyue12392** [comment]
@gfyoung  Tests added

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -1511,6 +1517,21 @@ cpdef bint is_integer_array(ndarray values):
     return validator.validate(values)
 
 
+cdef class IntegerNaValidator(Validator):
+    cdef inline bint is_value_typed(self, object value) except -1:
+        return util.is_integer_object(value) or util.is_nan(value)
+
+    cdef inline bint is_array_typed(self) except -1:
```
Comment (jreback) [inline pandas/_libs/lib.pyx]: I am not sure you need this one, is it actually hit?

File: pandas/_libs/lib.pyx [hunk 2]
```diff
@@ -1511,6 +1517,21 @@ cpdef bint is_integer_array(ndarray values):
     return validator.validate(values)
 
 
+cdef class IntegerNaValidator(Validator):
+    cdef inline bint is_value_typed(self, object value) except -1:
+        return util.is_integer_object(value) or util.is_nan(value)
```
Comment (jreback) [inline pandas/_libs/lib.pyx]: I think you need a is_valid_null as well (e.g. see how PeriodValidator is done)

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -472,7 +472,7 @@ def __new__(
                         pass
 
                     return Index(subarr, copy=copy, dtype=object, name=name)
-                elif inferred in ["floating", "mixed-integer-float"]:
+                elif inferred in ["floating", "mixed-integer-float", "integer-na"]:
```
Comment (jreback) [inline pandas/core/indexes/base.py]: ccan you add a TODO that we can return a Integer Array at some point

File: pandas/core/indexes/base.py [hunk 4]
```diff
@@ -3196,7 +3197,10 @@ def is_int(v):
                     self.get_loc(stop)
                 is_positional = False
         except KeyError:
-            if self.inferred_type == "mixed-integer-float":
+            if (
```
Comment (jreback) [inline pandas/core/indexes/base.py]: self.inferred_type in [....]

File: pandas/tests/dtypes/test_inference.py [hunk 5]
```diff
@@ -577,6 +577,16 @@ def test_integers(self):
         result = lib.infer_dtype(arr, skipna=True)
         assert result == "integer"
 
+    # GH 27392
```
Comment (jreback) [inline pandas/tests/dtypes/test_inference.py]: can you move to inside the definition

File: pandas/tests/dtypes/test_inference.py [hunk 6]
```diff
@@ -577,6 +577,16 @@ def test_integers(self):
...
+    def test_integer_na(self):
```
Comment (jreback) [inline pandas/tests/dtypes/test_inference.py]: can you parameterize over skipna?

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. you can add a note on api-breaking changes that ``infer_dtype`` will now return integer-na for mixed-floats

File: pandas/tests/dtypes/test_inference.py (hunk 6 above)
Comment (jiangyue12392) [reply pandas/tests/dtypes/test_inference.py]: Parameterization done

File: pandas/tests/dtypes/test_inference.py (hunk 5 above)
Comment (jiangyue12392) [reply pandas/tests/dtypes/test_inference.py]: Updated

File: pandas/core/indexes/base.py (hunk 4 above)
Comment (jiangyue12392) [reply pandas/core/indexes/base.py]: Updated

File: pandas/core/indexes/base.py (hunk 3 above)
Comment (jiangyue12392) [reply pandas/core/indexes/base.py]: TODO added

File: pandas/_libs/lib.pyx (hunk 2 above)
Comment (jiangyue12392) [reply pandas/_libs/lib.pyx]: I changed to `util.is_nan(value) and util.is_float_object(value)` instead

File: pandas/_libs/lib.pyx (hunk 1 above)
Comment (jiangyue12392) [reply pandas/_libs/lib.pyx]: Removed

File: doc/source/whatsnew/v0.25.0.rst [hunk 7]
```diff
@@ -825,6 +825,7 @@ Other API changes
 - :meth:`ExtensionArray.argsort` places NA values at the end of the sorted array. (:issue:`21801`)
 - :meth:`DataFrame.to_hdf` and :meth:`Series.to_hdf` will now raise a ``NotImplementedError`` when saving a :class:`MultiIndex` with extention data types for a ``fixed`` format. (:issue:`7775`)
 - Passing duplicate ``names`` in :meth:`read_csv` will now raise a ``ValueError`` (:issue:`17346`)
+- :meth:`infer_type` will now return integer-na for integer and np.nan mix (:issue:`27283`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you put integer-na in quotes here; move to 1.0

**Review by jreback — CHANGES_REQUESTED** [review]
whatsnew change, otherwise lgtm.

File: doc/source/whatsnew/v0.25.0.rst (hunk 7 above)
Comment (jiangyue12392) [reply doc/source/whatsnew/v0.25.0.rst]: Updated and moved

File: doc/source/whatsnew/v1.0.0.rst [hunk 8]
```diff
@@ -45,6 +45,7 @@ Backwards incompatible API changes
 Other API changes
 ^^^^^^^^^^^^^^^^^
 
+- :meth:`infer_type` will now return "integer-na" for integer and np.nan mix (:issue:`27283`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: I think this needs to be ``:meth:`pandas.api.types.infer_dtype` ``

**Review by jreback — APPROVED** [review]
lgtm. small change in the whatsnew, ping on green.

**Comment by jiangyue12392** [comment]
@jreback Green now

File: doc/source/whatsnew/v1.0.0.rst (hunk 8 above)
Comment (jiangyue12392) [reply doc/source/whatsnew/v1.0.0.rst]: Updated

File: doc/source/whatsnew/v1.0.0.rst [hunk 9]
```diff
@@ -45,6 +45,7 @@ Backwards incompatible API changes
 Other API changes
 ^^^^^^^^^^^^^^^^^
 
+- :meth:`pandas.api.types.infer_type` will now return "integer-na" for integer and np.nan mix (:issue:`27283`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: infer_type -> infer_dtype

**Comment by jreback** [comment]
@gfyoung ok here?

File: doc/source/whatsnew/v1.0.0.rst (hunk 9 above)
Comment (jiangyue12392) [reply doc/source/whatsnew/v1.0.0.rst]: Sure, updated

File: doc/source/whatsnew/v1.0.0.rst [hunk 10]
```diff
@@ -45,6 +45,7 @@ Backwards incompatible API changes
 Other API changes
 ^^^^^^^^^^^^^^^^^
 
+- :meth:`pandas.api.types.infer_dtype` will now return "integer-na" for integer and np.nan mix (:issue:`27283`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v1.0.0.rst]: Double back-ticks around `np.nan`

File: pandas/tests/dtypes/test_inference.py [hunk 11]
```diff
@@ -577,6 +577,18 @@ def test_integers(self):
...
+    def test_integer_na(self, skipna):
+        # GH 27392
+        expected = "integer" if skipna else "integer-na"
+        arr = np.array([1, 2, np.nan, np.nan, 3], dtype="O")
+        result = lib.infer_dtype(arr, skipna=skipna)
+        assert result == expected
+
+        arr = np.array([1, 2, 3, np.int64(4), np.int32(5), np.nan], dtype="O")
+        result = lib.infer_dtype(arr, skipna=skipna)
+        assert result == expected
```
Comment (gfyoung) [inline pandas/tests/dtypes/test_inference.py]: Can we parameterize `arr` as well here?

File: doc/source/whatsnew/v1.0.0.rst (hunk 10 above)
Comment (jiangyue12392) [reply doc/source/whatsnew/v1.0.0.rst]: Added

File: pandas/tests/dtypes/test_inference.py (hunk 11 above)
Comment (jiangyue12392) [reply pandas/tests/dtypes/test_inference.py]: Parameterization done

File: pandas/tests/dtypes/test_inference.py [hunk 12]
```diff
@@ -577,6 +577,28 @@ def test_integers(self):
         result = lib.infer_dtype(arr, skipna=True)
         assert result == "integer"
 
+    @pytest.mark.parametrize(
+        "arr, skipna, expected",
```
Comment (gfyoung) [inline pandas/tests/dtypes/test_inference.py]: You don't need to parameterize `expected` here.  It's `integer-na` when `skipa` is `False` and `integer` when `skipna` is `True`.

Comment (jiangyue12392) [reply pandas/tests/dtypes/test_inference.py]: Ok, sure, updated

File: pandas/tests/dtypes/test_inference.py [hunk 13]
```diff
@@ -577,6 +577,21 @@ def test_integers(self):
         result = lib.infer_dtype(arr, skipna=True)
         assert result == "integer"
 
+    @pytest.mark.parametrize(
+        "arr, skipna",
+        [
+            (np.array([1, 2, np.nan, np.nan, 3], dtype="O"), False),
```
Comment (gfyoung) [inline pandas/tests/dtypes/test_inference.py]: Don't forget to drop the `expected` fixture from your parameterization setup!

Comment (jiangyue12392) [reply pandas/tests/dtypes/test_inference.py]: Oh, right, sorry, sure

**Review by gfyoung — APPROVED** [review]
Nice!  @jreback back over to you

**Comment by jreback** [comment]
thanks @jiangyue12392 keep em coming! you are tackling non-trivial issues and making very nice PRs!

**Pull request merged**